streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorTest.java

```diff
@@ -118,49 +134,68 @@ public void intermediateSuppressionShouldThrow() {
     }
 
 
-    @Test
-    public void finalResultsSuppressionShouldThrow() {
-        final KTableSuppressProcessor<Windowed<String>, Long> processor =
-            new KTableSuppressProcessor<>(finalResults(ofMillis(1)));
+    @Test(expected = KTableSuppressProcessor.NotImplementedException.class)
+    public void finalResultsSuppressionShouldBufferAndEmitLater() {
```

Comment:
What's the difference between this test and the `finalResultsWith0GraceBeforeWindowEndShouldBufferAndEmitLater` below? Besides `grace` they seems exactly the same.

In that case, could we just merge these two, to randomly choose one as the grace period, and with a significant amount of likelihood that the random number is 0?

Reply:
They are similar, but targeting different behaviors:
* `finalResultsSuppressionShouldBufferAndEmitLater` is the "main" unit test: verifying that the buffer saves up all the data and emits at the grace period expiration.
* `finalResultsWith0GraceBeforeWindowEndShouldBufferAndEmitLater` is somewhat of a "regression" test. I added it during implementation because I initially failed to account for this behavior: even when the grace period is 0, the final results buffer should wait until the end of the window before emitting (instead of immediately emitting). I think I can name the method better and add a comment to make this clear.

As far as randomized testing goes, I'd like to experiment with a fuzzing framework in the future, rather than hacking out a half-baked one right now. There are some important features, such as a mechanism to reproduce failing tests, that are necessary to make it a "randomized test", rather than a "flaky test".

For now, I'd rather just stick with hand-written scenarios.

Reply:
Makes sense.